### PR TITLE
feat: idempotent-copy file

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. idempotent-copy file
2. add `force` option in `copy into` SQL.
3. add `truncate_table` in catalog to truncate all the copied files info in meta service.
4. add test for idempotent-copy and force copy.

Fixes https://github.com/datafuselabs/databend/issues/7550, https://github.com/datafuselabs/databend/issues/7561

### CODE REVIEW

File: src/query/service/src/interpreters/interpreter_copy_v2.rs [hunk 1]
```diff
@@ -49,6 +54,95 @@ impl CopyInterpreterV2 {
...
+            files: files.to_owned(),
+        };
+        let mut file_map = BTreeMap::new();
+
+        if !force {
+            let resp = catalog.get_table_copied_file_info(req).await?;
+            for file in files.iter() {
+                let stage_file = list_files(&self.ctx, &table_info.stage_info, file, "").await?;
+                if stage_file.is_empty() {
+                    continue;
```
Comment (BohuTANG) [inline src/query/service/src/interpreters/interpreter_copy_v2.rs:83]: Should add a `warn!` log to remind the file is skipped.

File: src/query/service/src/interpreters/interpreter_copy_v2.rs [hunk 2]
```diff
@@ -49,6 +54,95 @@ impl CopyInterpreterV2 {
...
+                }
+
+                let stage_file = &stage_file[0];
+
+                if let Some(file_info) = resp.file_info.get(file) {
+                    // No need to copy the file again if etag is_some and match.
+                    if stage_file.etag.is_some() && stage_file.etag == file_info.etag {
+                        continue;
```
Comment (BohuTANG) [inline src/query/service/src/interpreters/interpreter_copy_v2.rs:91]: Should add a warn! log to remind the file&meta is skipped.

File: src/query/service/src/interpreters/interpreter_copy_v2.rs [hunk 3]
```diff
@@ -49,6 +54,95 @@ impl CopyInterpreterV2 {
...
+                    }
+
+                    if file_info.content_length == stage_file.size
+                        && file_info.last_modified == Some(stage_file.last_modified)
+                    {
```
Comment (BohuTANG) [inline src/query/service/src/interpreters/interpreter_copy_v2.rs:96]: When do we fall in this case?

Comment (BohuTANG) [reply src/query/service/src/interpreters/interpreter_copy_v2.rs:96]: If the ETag is the same, does the check is not need?

File: src/query/service/src/interpreters/interpreter_copy_v2.rs [hunk 4]
```diff
@@ -320,14 +415,59 @@ impl Interpreter for CopyInterpreterV2 {
...
+                                &files,
+                            )
+                            .await?;
+
+                        tracing::info!(
+                            "matched copy unduplicate files: {:?}",
+                            &copy_stage_files.keys(),
+                        );
+
+                        if copy_stage_files.is_empty() {
```
Comment (BohuTANG) [inline src/query/service/src/interpreters/interpreter_copy_v2.rs:437]: It looks better to check the empty in `copy_files_to_table` function.

**Comment by BohuTANG** [comment]
@mergify update

File: src/query/service/src/interpreters/interpreter_copy_v2.rs (hunk 3 above)
Comment (lichuang) [reply src/query/service/src/interpreters/interpreter_copy_v2.rs:98]: > If the ETag is the same, does the check is not needed?

yes, now the logic is:

1. compare etag, if etag is_some and matched, ignore copy this file.
2. compare content-length and last_modified, if matched, ignore copy this file.

Comment (BohuTANG) [reply src/query/service/src/interpreters/interpreter_copy_v2.rs:98]: Do not get the point:
If the etag is the same why need to check content_length&size again?

File: src/query/service/src/interpreters/interpreter_copy_v2.rs [hunk 5]
```diff
@@ -49,6 +54,108 @@ impl CopyInterpreterV2 {
...
+                file_map.insert(file.clone(), TableCopiedFileInfo {
+                    etag: stage_file.etag.clone(),
+                    content_length: stage_file.size,
+                    last_modified: Some(stage_file.last_modified),
+                });
+            }
+        } else {
+            // if force is true, copy all the file.
+            for file in files.iter() {
+                let stage_file = list_files(&self.ctx, &table_info.stage_info, file, "").await?;
```
Comment (Xuanwo) [inline src/query/service/src/interpreters/interpreter_copy_v2.rs:123]: Don't use `list_files` to get the file's metadata. We can implement a function like the following:

```rust
async fn stat_file(ctx: &Arc<QueryContext>, stage: &UserStageInfo, path: &str) -> Result<StageFile>`
```

In which we can use `op.object(path).metadata().await`.

**Review by BohuTANG — APPROVED** [review]

**Comment by BohuTANG** [comment]
@soyeric128 
Please update the copy command, we introduced a new option: FORCE

**Pull request merged**